File: ch04/python/exercises.py
"""
Chapter 04: Exercises
"""

def quadratic_greatest_number(list):
    counter = 0

    for i in list:
        # Assume for now that i is the greatest:
        is_i_val_the_greatest = True

        for j in list:
            counter += 1
            # If we find another value that is greater than i,
            # i is not the greatest:
            if j > i:
                is_i_val_the_greatest = False

        # If, by the time we checked all the other numbers, i
        # is still the greatest, it means that i is the greatest number:
        if is_i_val_the_greatest:
            print(f"Greatest Number: {i}")
            print(f"Steps: {counter}")
            return counter
        

print("******** O(N^2): Greatest Number ********")
number_list = [3, 1, 6, 8, 2, 9]
steps = quadratic_greatest_number(number_list)
assert steps == 36


def linear_greatest_number(list):
    greatest_number = list[0]
    counter = 0

    for i in range(len(list)):
        counter += 1

        if list[i] > greatest_number:
            greatest_number = list[i]

    print(f"Greatest Number: {greatest_number}")
    print(f"Steps: {counter}")
    return counter


print("******** O(N): Greatest Number ********")
number_list = [3, 1, 6, 8, 2, 9]
steps = linear_greatest_number(number_list)
assert steps == 6

File: ch04/go/exercises.go
package ch04

import "fmt"

func QuadraticGreatestNumber(list [6]int) int {
	counter := 0

	for _, i := range list {
		isIValTheGreatest := true

		for _, j := range list {
			counter++

			if j > i {
				isIValTheGreatest = false
			}
		}

		if isIValTheGreatest {
			fmt.Println("Greatest Number: ", i)
			fmt.Println("Steps: ", counter)
			return counter
		}
	}

	return counter
}

func LinearGreatestNumber(list [6]int) int {
	greatestNumber := list[0]
	counter := 0

	for i := range list {
		counter++

		if list[i] > greatestNumber {
			greatestNumber = list[i]
		}
	}

	fmt.Println("Greatest Number: ", greatestNumber)
	fmt.Println("Steps: ", counter)
	return counter
}
